Node Hangs

I observed that my node is hanging. Not sure why. Last few log messages:

```
2024-12-05T21:53:23.854059211Z  INFO ThreadId(02) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:23.854096191Z  WARN ThreadId(02) neptune_cash::peer_loop: Received 1 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:23.854124811Z  WARN ThreadId(02) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 0.320761143 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:23.854308201Z  INFO ThreadId(02) neptune_cash::peer_loop: Got new block from peer [::ffff:107.181.240.234]:41672, height 2210, mined 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:23.854733721Z  INFO ThreadId(02) neptune_cash::mine_loop: Miner task received beta block height 2210
2024-12-05T21:53:23.942761497Z  INFO ThreadId(03) neptune_cash::peer_loop: Block with height 2209 is valid. mined: 2024-12-05T21:32:56.300+00:00
2024-12-05T21:53:24.177705063Z  INFO ThreadId(02) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:24.261931122Z  INFO ThreadId(03) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:24.49263834Z  INFO ThreadId(02) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:24.57640586Z  INFO ThreadId(03) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:24.811175735Z  INFO ThreadId(02) neptune_cash::peer_loop: Block with height 2210 is valid. mined: 2024-12-05T21:53:17.487+00:00
2024-12-05T21:53:24.811267855Z  WARN ThreadId(03) neptune_cash::peer_loop: Received 1 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:24.811291875Z  WARN ThreadId(03) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 1.278402247 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:24.811448405Z  WARN ThreadId(03) neptune_cash::peer_loop: Received 1 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:24.811475485Z  WARN ThreadId(03) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 1.278658637 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:24.811571325Z  WARN ThreadId(03) neptune_cash::peer_loop: Received 10 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:24.811692815Z  WARN ThreadId(03) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 3.857305739 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:24.811788645Z  WARN ThreadId(02) neptune_cash::peer_loop: Received 1 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:24.811810895Z  WARN ThreadId(03) neptune_cash::peer_loop: Received 1 blocks from peer but incoming blocks are less canonical than current tip.
2024-12-05T21:53:24.811843046Z  WARN ThreadId(02) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 1.278586039 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:24.811851025Z  WARN ThreadId(03) neptune_cash: executed handle_peer_message()::PeerMessage::Block in 0.957542794 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:532:17
2024-12-05T21:53:24.812810554Z  WARN ThreadId(01) neptune_cash: executed run()::select::mp_resync_timer in 0.001559499 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/main_loop.rs:1468:21
2024-12-05T21:53:24.994879006Z  WARN ThreadId(03) neptune_cash: executed handle_peer_message()::PeerMessage::BlockRequestByHeight in 0.182773032 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:850:17
2024-12-05T21:53:25.11751951Z  WARN ThreadId(02) neptune_cash: executed handle_peer_message()::PeerMessage::BlockRequestByHash in 0.003222898 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:826:17
2024-12-05T21:54:49.019940768Z  INFO ThreadId(01) neptune_cash::main_loop: Attempting to reconnect to peer with lost connection: 139.162.193.206:9798
2024-12-05T21:54:49.020016008Z  INFO ThreadId(01) neptune_cash::main_loop: Performing peer discovery
2024-12-05T21:54:49.020037568Z  INFO ThreadId(01) neptune_cash::main_loop: Connecting to peer [::ffff:79.157.213.217]:9798 with distance 2
2024-12-05T21:54:49.020395938Z  INFO ThreadId(03) neptune_cash::connect_to_peers: Established outgoing TCP connection with 139.162.193.206:9798
2024-12-05T21:54:49.020673748Z  INFO ThreadId(02) neptune_cash::connect_to_peers: Established incoming TCP connection with [::ffff:139.162.193.206]:48210
2024-12-05T21:54:49.020956617Z  WARN ThreadId(02) neptune_cash::connect_to_peers: Incoming connection refused: SelfConnect
2024-12-05T21:54:49.021014877Z ERROR ThreadId(02) neptune_cash::main_loop: Got error: Refusing incoming connection. Reason: SelfConnect
2024-12-05T21:54:49.021043447Z ERROR ThreadId(03) neptune_cash::connect_to_peers: An error occurred: Outgoing connection attempt refused. Reason: SelfConnect. Connection closing
2024-12-05T21:54:49.021066747Z  INFO ThreadId(03) neptune_cash::connect_to_peers: Connection closing
2024-12-05T21:55:02.601627685Z  WARN ThreadId(02) neptune_cash::connect_to_peers: Failed to establish connection: Connection timed out (os error 110)
2024-12-05T21:55:02.601689695Z  INFO ThreadId(02) neptune_cash::connect_to_peers: Connection closing
2024-12-05T21:55:04.838148482Z  INFO ThreadId(02) neptune_cash::peer_loop: Got block proposal from peer.
2024-12-05T21:55:05.152632797Z  INFO ThreadId(02) neptune_cash::peer_loop: Rewarding peer ::ffff:66.248.205.33 for NewBlockProposal
2024-12-05T21:55:05.152686637Z  WARN ThreadId(02) neptune_cash: executed handle_peer_message()::PeerMessage::BlockProposal in 0.314540425 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:1128:17
2024-12-05T21:55:05.152760117Z  INFO ThreadId(01) neptune_cash::main_loop: Received new favorable block proposal for mining operation.
2024-12-05T21:55:05.153318377Z  INFO ThreadId(03) neptune_cash::mine_loop: Miner received message about new block proposal for guessing.
2024-12-05T21:55:05.154226766Z  INFO ThreadId(572) neptune_cash::mine_loop: Mining on block with 1 outputs and difficulty 704690060. Attempting to find block with height 2211 with digest less than target: 05169561069146539959,10291194454790015845,06096537823767786088,16985605236488289992,00000000026177102695
2024-12-05T21:55:05.224629293Z  INFO ThreadId(02) neptune_cash::peer_loop: Got block proposal from peer.
2024-12-05T21:55:05.224682013Z  WARN ThreadId(02) neptune_cash::peer_loop: Rejecting new block proposal:
Insufficient fee. Proposal was 64;
current fee is: Some(NeptuneCoins(256000000000000005090463958695936))
2024-12-05T21:55:05.225672992Z  WARN ThreadId(02) neptune_cash::peer_loop: Punishing peer 51.15.139.238 for NonFavorableBlockProposal
2024-12-05T21:55:05.225688382Z  WARN ThreadId(02) neptune_cash: executed handle_peer_message()::PeerMessage::BlockProposal in 0.001061049 secs.  exceeds slow fn threshold of 0.001 secs.  location: /home/runner/work/neptune-core/neptune-core/src/peer_loop.rs:1128:17
2024-12-05T21:56:31.651265531Z  INFO ThreadId(02) neptune_cash::peer_loop: Got new block from peer 51.15.139.238:9798, height 2211, mined 2024-12-05T21:56:29.441+00:00
2024-12-05T21:56:32.264116975Z  INFO ThreadId(02) neptune_cash::peer_loop: Got new block from peer 151.115.78.81:9798, height 2211, mined 2024-12-05T21:56:29.441+00:00
```

The dashboard fails to make a connection. Ctrl-C does not work. I ran it with the following command: `neptune-core --max-num-peers 25 --bootstrap --guess --sleepy-guessing --peers 51.15.139.238:9798 --peers 139.162.193.206:9798`. After reboot it seems to be syncing okay but I am noticing this worrying error popping up a bunch of times:

```
2024-12-10T14:25:48.367618193Z ERROR ThreadId(01) neptune_cash::main_loop: Failed to send pause message to miner thread:                                                                                                             
no available capacity                                                                                                                                                                                                                
```

Let me note something here for the record, just in case it's relevant later.

I'm changing this function in `BlockProgram`:
```
    pub(crate) fn verify(block_body: &BlockBody, appendix: &BlockAppendix, proof: &Proof) -> bool {
        let claim = Self::claim(block_body, appendix);
        triton_vm::verify(Stark::default(), &claim, proof)
    }
```
into something resembling this:
```
    pub(crate) async fn verify(
        block_body: &BlockBody,
        appendix: &BlockAppendix,
        proof: &Proof,
    ) -> bool {
        let claim = Self::claim(block_body, appendix);
        task::spawn_blocking(|| triton_vm::verify(Stark::default(), &claim, proof))
            .await
            .expect("should be able to verify block proof in new tokio task")
    }
```
Note the `async` and the `spawn_blocking`.

The point is that the second one doesn't compile. The compiler generates this error:
```
error[E0373]: closure may outlive the current function, but it borrows `claim`, which is owned by the current function
  --> src/models/blockchain/block/validity/block_program.rs:47:30
   |
47 |         task::spawn_blocking(|| triton_vm::verify(Stark::default(), &claim, proof))
   |                              ^^ may outlive borrowed value `claim`   ----- `claim` is borrowed here
   |
note: function requires argument type to outlive `'static`
  --> src/models/blockchain/block/validity/block_program.rs:47:9
   |
47 |         task::spawn_blocking(|| triton_vm::verify(Stark::default(), &claim, proof))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `claim` (and any other referenced variables), use the `move` keyword
   |
47 |         task::spawn_blocking(move || triton_vm::verify(Stark::default(), &claim, proof))
   |                              ++++

error[E0521]: borrowed data escapes outside of associated function
  --> src/models/blockchain/block/validity/block_program.rs:47:9
   |
44 |         proof: &Proof,
   |         -----  - let's call the lifetime of this reference `'1`
   |         |
   |         `proof` is a reference that is only valid in the associated function body
...
47 |         task::spawn_blocking(|| triton_vm::verify(Stark::default(), &claim, proof))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         `proof` escapes the associated function body here
   |         argument requires that `'1` must outlive `'static`
```

There are two lifetime issues, one for `claim` and one for `proof`. The `claim` is owned by the current function and that can be `move`d without any problems. The proof on the other hand is only passed by reference. I fixed it by `clone`ing it (and binding it to a variable owned by the function), but it makes me wonder whether this lifetime issue is at play in generating the crash. We'll have more information if the modified code produces the same error.

--- a/src/config_models/cli_args.rs
+++ b/src/config_models/cli_args.rs
@@ -113,6 +113,11 @@ pub struct Args {
     #[clap(long)]
     pub sleepy_guessing: bool,
 
+    /// Set the number of threads to use while guessing. When no value is set,
+    /// the number is set to the number of available cores.
+    #[clap(long)]
+    pub guesser_threads: Option<usize>,
+
     /// Determines the fraction of the transaction fee consumed by this node as
     /// a reward either for upgrading a `ProofCollection` to `SingleProof`, or
     /// for merging two `SingleProof`s.

--- a/src/mine_loop.rs
+++ b/src/mine_loop.rs
@@ -98,13 +98,13 @@ async fn compose_block(
 }
 
 /// Attempt to mine a valid block for the network
-#[allow(clippy::too_many_arguments)]
 async fn guess_nonce(
     block: Block,
     previous_block_header: BlockHeader,
     sender: oneshot::Sender<NewBlockFound>,
     composer_utxos: Vec<ExpectedUtxo>,
     sleepy_guessing: bool,
+    num_guesser_threads: Option<usize>,
     target_block_interval: Option<Timestamp>,
 ) {
     // We wrap mining loop with spawn_blocking() because it is a
@@ -126,6 +126,7 @@ async fn guess_nonce(
             sender,
             composer_utxos,
             sleepy_guessing,
+            num_guesser_threads,
             target_block_interval,
         )
     })
@@ -151,18 +152,21 @@ fn guess_worker(
     sender: oneshot::Sender<NewBlockFound>,
     composer_utxos: Vec<ExpectedUtxo>,
     sleepy_guessing: bool,
+    num_guesser_threads: Option<usize>,
     target_block_interval: Option<Timestamp>,
 ) {
     // This must match the rules in `[Block::has_proof_of_work]`.
     let prev_difficulty = previous_block_header.difficulty;
     let threshold = prev_difficulty.target();
     let input_block_height = block.kernel.header.height;
+    let threads_to_use = num_guesser_threads.unwrap_or_else(rayon::current_num_threads);
     info!(
-        "Mining on block with {} outputs and difficulty {}. Attempting to find block with height {} with digest less than target: {}",
+        "Guessing with {} threads on block {} with {} outputs and difficulty {}. Target: {}",
+        threads_to_use,
+        block.header().height,
         block.body().transaction_kernel.outputs.len(),
         previous_block_header.difficulty,
-        block.header().height,
-        threshold
+        threshold.to_hex()
     );
 
     // note: this article discusses rayon strategies for mining.
@@ -172,9 +176,7 @@ fn guess_worker(
     // see:  https://docs.rs/rayon/latest/rayon/fn.max_num_threads.html
     let block_header_template = block.header().to_owned();
     let (block_body_mast_hash_digest, appendix_digest) = precalculate_mast_leafs(&block);
-    let rayon_threads_available = rayon::current_num_threads();
 
-    let threads_to_use = rayon_threads_available;
     let pool = ThreadPoolBuilder::new()
         .num_threads(threads_to_use)
         .build()
@@ -656,6 +658,7 @@ pub(crate) async fn mine(
                 guesser_tx,
                 composer_utxos,
                 cli_args.sleepy_guessing,
+                cli_args.guesser_threads,
                 None, // using default TARGET_BLOCK_INTERVAL
             );
 
@@ -1333,6 +1336,7 @@ pub(crate) mod mine_loop_tests {
         );
 
         let sleepy_guessing = false;
+        let num_guesser_threads = None;
 
         guess_worker(
             block,
@@ -1340,6 +1344,7 @@ pub(crate) mod mine_loop_tests {
             worker_task_tx,
             coinbase_utxo_info,
             sleepy_guessing,
+            num_guesser_threads,
             None,
         );
 
@@ -1407,6 +1412,7 @@ pub(crate) mod mine_loop_tests {
         assert_eq!(ten_seconds_ago, initial_header_timestamp);
 
         let sleepy_guessing = false;
+        let num_guesser_threads = None;
 
         guess_worker(
             template,
@@ -1414,6 +1420,7 @@ pub(crate) mod mine_loop_tests {
             worker_task_tx,
             coinbase_utxo_info,
             sleepy_guessing,
+            num_guesser_threads,
             None,
         );
 
@@ -1492,6 +1499,7 @@ pub(crate) mod mine_loop_tests {
 
         // set initial difficulty in accordance with own hash rate
         let sleepy_guessing = false;
+        let num_guesser_threads = None;
         let num_outputs = 0;
         let hash_rate =
             estimate_own_hash_rate(Some(target_block_interval), sleepy_guessing, num_outputs).await;
@@ -1566,6 +1574,7 @@ pub(crate) mod mine_loop_tests {
                 worker_task_tx,
                 composer_utxos,
                 sleepy_guessing,
+                num_guesser_threads,
                 Some(target_block_interval),
             );
 

--- a/src/mine_loop.rs
+++ b/src/mine_loop.rs
@@ -1,6 +1,5 @@
 pub(crate) mod composer_parameters;
 
-use std::cmp;
 use std::cmp::max;
 use std::time::Duration;
 
@@ -174,12 +173,8 @@ fn guess_worker(
     let block_header_template = block.header().to_owned();
     let (block_body_mast_hash_digest, appendix_digest) = precalculate_mast_leafs(&block);
     let rayon_threads_available = rayon::current_num_threads();
-    if rayon_threads_available < 2 {
-        error!("Cannot guess when less than two threads are available to rayon");
-        return;
-    }
 
-    let threads_to_use = cmp::max(1, rayon_threads_available.saturating_sub(2));
+    let threads_to_use = rayon_threads_available;
     let pool = ThreadPoolBuilder::new()
         .num_threads(threads_to_use)
         .build()

--- a/src/mine_loop.rs
+++ b/src/mine_loop.rs
@@ -1,5 +1,6 @@
 pub(crate) mod composer_parameters;
 
+use std::cmp;
 use std::cmp::max;
 use std::time::Duration;
 
@@ -16,9 +17,9 @@ use rand::rngs::StdRng;
 use rand::Rng;
 use rand::SeedableRng;
 use rayon::iter::ParallelIterator;
+use rayon::ThreadPoolBuilder;
 use tasm_lib::prelude::Tip5;
 use tasm_lib::triton_vm::prelude::BFieldCodec;
-use tasm_lib::twenty_first::prelude::CpuParallel;
 use tasm_lib::twenty_first::prelude::MerkleTreeMaker;
 use tokio::select;
 use tokio::sync::mpsc;
@@ -49,6 +50,7 @@ use crate::models::state::wallet::transaction_output::TxOutputList;
 use crate::models::state::GlobalState;
 use crate::models::state::GlobalStateLock;
 use crate::prelude::twenty_first;
+use crate::twenty_first::util_types::merkle_tree::CpuParallel;
 use crate::COMPOSITION_FAILED_EXIT_CODE;
 
 async fn compose_block(
@@ -171,33 +173,43 @@ fn guess_worker(
     // see:  https://docs.rs/rayon/latest/rayon/fn.max_num_threads.html
     let block_header_template = block.header().to_owned();
     let (block_body_mast_hash_digest, appendix_digest) = precalculate_mast_leafs(&block);
-    let guess_result = rayon::iter::repeat(0)
-        .map_init(
-            || {
-                (
-                    &previous_block_header,
-                    block_header_template.clone(),
-                    rand::thread_rng(),
-                )
-            },
-            |(prev_header, block_header_templ, rng), _x| {
-                guess_nonce_iteration(
-                    block_body_mast_hash_digest,
-                    appendix_digest,
-                    block_header_templ.to_owned(),
-                    prev_header,
-                    &sender,
-                    DifficultyInfo {
-                        target_block_interval,
-                        threshold,
-                    },
-                    sleepy_guessing,
-                    rng,
-                )
-            },
-        )
-        .find_any(|r| !r.block_not_found())
+    let rayon_threads_available = rayon::current_num_threads();
+    let threads_to_use = cmp::max(1, rayon_threads_available.saturating_sub(2));
+    let pool = ThreadPoolBuilder::new()
+        .num_threads(threads_to_use)
+        .build()
         .unwrap();
+    let guess_result = pool.install(|| {
+        rayon::iter::repeat(0)
+            .map_init(
+                || {
+                    (
+                        block_body_mast_hash_digest,
+                        appendix_digest,
+                        &block_header_template,
+                        rand::thread_rng(),
+                    )
+                },
+                |(block_body_mast_hash_digest_, appendix_digest_, block_header_template_, rng),
+                 _x| {
+                    guess_nonce_iteration(
+                        *block_body_mast_hash_digest_,
+                        *appendix_digest_,
+                        block_header_template_.clone(),
+                        &previous_block_header,
+                        &sender,
+                        DifficultyInfo {
+                            target_block_interval,
+                            threshold,
+                        },
+                        sleepy_guessing,
+                        rng,
+                    )
+                },
+            )
+            .find_any(|r| !r.block_not_found())
+            .unwrap()
+    });
 
     let (nonce_preimage, timestamp, difficulty) = match guess_result {
         GuessNonceResult::Cancelled => {

--- a/src/mine_loop.rs
+++ b/src/mine_loop.rs
@@ -174,6 +174,11 @@ fn guess_worker(
     let block_header_template = block.header().to_owned();
     let (block_body_mast_hash_digest, appendix_digest) = precalculate_mast_leafs(&block);
     let rayon_threads_available = rayon::current_num_threads();
+    if rayon_threads_available < 2 {
+        error!("Cannot guess when less than two threads are available to rayon");
+        return;
+    }
+
     let threads_to_use = cmp::max(1, rayon_threads_available.saturating_sub(2));
     let pool = ThreadPoolBuilder::new()
         .num_threads(threads_to_use)

